Support empty string handling for date and time datatypes

manisha-deshpande opened the pull request:
### Overview
Add TSQL-compatible handling of empty string ('') for date and time types, defaulting to '1900-01-01' for date and time, similar to datetime, datetime2, datetimeoffset, smalldatetime.

Task: BABEL-3433

### Description

In SQL Server, when an empty string ('') is cast to `date` or `time` datatype, it should be return a default value - '1900-01-01' for date and '00:00:00.0000000' for time. Currently, Babelfish raises a syntax error for these cases. This change improves TSQL compatibility by matching SQL Server's behavior for empty string handling in date and time datatypes, while maintaining existing behavior for non-empty inputs.

```
1> create table bug(dt date null, tm time(3) null);
2> insert into bug values ('', '');
go
Msg 33557097, Level 16, State 1, Server BABELFISH, Line 1
invalid input syntax for type date: ""
Msg 33557097, Level 16, State 1, Server BABELFISH, Line 1
invalid input syntax for type time: ""
```
 
### Changes:
1. __Default Value initialization helper functions:__
- Add helper functions `DateADT initializeToDefaultDate(void);` and `TimeADT initializeToDefaultTime(int32 typmod);` to initialize the default date and time variables.

2. __Whitespace detection function:__
 - Added `isEmptyOrWhitespace()` to safely check if a string is empty or consists only of whitespace characters
 - Implemented with proper boundary checking to ensure memory safety
 - Applied consistently across all datetime-related parsing functions


3. __Conversion path improvements:__
 - Modified existing functions to handle empty and whitespace-only inputs:
   - `varchar2date()` and `varchar2time()` - Handle conversions from varchar types
   - Updated `tsql_coerce_string_literal_hook()` - Handle direct string literals
 - Implementation leverages timestamp2tm/tm2time for consistent PG date/time conversion


4. __Test coverage:__
 - Added comprehensive test suite date_time_empty_input_tests with prepare/verify/cleanup structure
 - Test cases include:
   - Direct casts with empty and whitespace strings
   - Table insertions with empty/whitespace values
   - Variable assignments and type conversions
   - View creation and computed columns
   - Integration with other datetime functions


### Future Work (Out of Scope): BABEL-5923

- Support for casts from empty char/nchar/text/ntext strings to date/time types
- Tab character handling for all datetime datatypes


### Issues Resolved

- BABEL-3433: Support for empty input string handling in date and time datatypes
- Extended handling for whitespace-only strings (' ') for other datetime datatypes


### Test Scenarios Covered [+ `BABEL_3433.sql`]
* **Use case based -**
```
CREATE TABLE date_testing(d DATE);
INSERT INTO date_testing VALUES('');
go
INSERT ...

create table time_testing ( ti TIME );
INSERT INTO time_testing VALUES('');
go
INSERT ...
```
 ** The insertion subsequently affects the existing FORMAT date function tests with various arguments for tables date_testing, time_testing.
```
select * from date_testing;
go
~~START~~
date
1900-01-01
...

select FORMAT(d, 'd','en-us') from date_testing;
GO
~~START~~
nvarchar
1/1/1900
...
```


* Other date functions 
```
select CONVERT(varchar(30), CAST('' AS date), 102);
go -- 1900.01.01
select TRY_CONVERT(varchar(30), CAST('' AS date), 102);
go -- 1900.01.01
select datepart(day, cast('' as date));
go -- 1
select datediff(year, CAST('1900-01-01'AS date), CAST(''AS date));
go -- 0
select DAY(CAST('' AS date));
go -- 1
select YEAR(CAST('' AS date));
go -- 1900

```

* Additional cast to other date types
```
select CAST(CAST('' AS date) AS datetime)
go -- datetime 1900-01-01 00:00:00.0
select CAST(CAST('' AS time) AS datetime)
go -- datetime 1900-01-01 00:00:00.0
select CAST(CAST('' AS date) AS datetime2);
go -- datetime2 1900-01-01 00:00:00.0000000
select CAST(CAST('' AS time) AS datetime2);
go -- datetime2 1900-01-01 00:00:00.0000000
select CAST(CAST('' AS time) AS smalldatetime);
go -- smalldatetime 1900-01-01 00:00:00.0
select CAST(CAST('' AS date) AS smalldatetime);
go -- smalldatetime 1900-01-01 00:00:00.0
select CAST(CAST('' AS date) AS datetimeoffset);
go -- datetimeoffset 1900-01-01 00:00:00.0000000 +00:00
select CAST(CAST('' AS time) AS datetimeoffset);
go -- datetimeoffset 1900-01-01 00:00:00.0000000 +00:00

```

* **Boundary conditions -**


* **Arbitrary inputs -**


* **Negative test cases -**


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

manisha-deshpande committed:
Support empty string handling for date and time datatypes

Add TSQL-compatible handling of empty string ('') for date and time
types, defaulting to '1900-01-01' for date and '00:00:00.0000000' for
time, similar to datetime, datetime2, datetimeoffset, smalldatetime.
Added empty string test cases in different scenarios including table
insert, direct cast, cast to other date types, and formatting functions.

Changes:
1. Add helper functions to initialize the default date and time variables
   and to check for strings with whitespaces.
2. Modify varchar2date() and varchar2time() to handle empty/whitespace
   casts from varchar types to date and time types.
3. Update tsql_coerce_string_literal_hook to handle empty/whitespace
   casts from string literals to date and time datatypes.
4. Default empty/whitespace strings to '1900-01-01' for date and
   '00:00:00.0000000' for time.
5. Updated existing date-time related test cases. Added more tests cases
   for wider test coverage.

Task: BABEL-3433
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

RIC06X reviewed (approved):
LGTM

Anikait143 commented:
Please add following Test Cases:
1. Tests with empty strings consisting of whitespapces/tabs/new lines.
2. Declare/Select statements
3. Insertion into table with date/time column
4. Casts from string datatypes (empty string) to date/time.
5. Test with Views

Anikait143 commented on contrib/babelfishpg_tsql/src/pltsql_coerce.c:
```diff
@@ -2566,11 +2568,40 @@ tsql_coerce_string_literal_hook(Oid targetTypeId,
+
+			newcon->constvalue = DateADTGetDatum(date);
+		}
+		else if (i == -1 && targetTypeId == TIMEOID)
```
Let's fix this for other Datetime datatypes as well (Datetime, Datetime2, smalldatetime, Datetimeoffset)

manisha-deshpande replied:
The added check handles empty string specifically for casts (targeting) TO date/time.

The other date types already check and handle empty strings within their respective `_in` functions, such as `datetimeoffset_in`. Furthermore, each have different return logic, for example the `tm2time/date2j` or `TimeADTGetDatum/DateADTGetDatum` used here would not work for `datetimeoffset` which invokes `AdjustDatetimeoffsetForTypmod` after `initializeToDefaultDatetime`. 

To implement the suggestion - moving everything under `case baseTypeId == DATEOID/TIMEOID`, we would need to write and invoke a generalized function with switch case covering all the datetypes, and move the existing empty string handling logic(s) out from their existing functions to the new one.

manisha-deshpande replied:
fixed

Anikait143 commented on test/JDBC/input/BABEL-3433.sql:
Nit: Please have proper descriptive file name, we don't use Jira numbers anymore.

Anikait143 commented on test/JDBC/input/BABEL-3433.sql:
Let's add this test file to latest upgrade schedule as well.

Anikait143 commented on contrib/babelfishpg_tsql/src/pltsql_coerce.c:
```diff
@@ -2566,11 +2568,40 @@ tsql_coerce_string_literal_hook(Oid targetTypeId,
 		}
 
-		if (i == -1)
+		if (i == -1 && targetTypeId == DATEOID)
```
We should use `baseTypeId` instead and also let's combine with the switch case below

manisha-deshpande replied:
fixed

Anikait143 commented on contrib/babelfishpg_common/src/varchar.c:
Can we handle cases for empty string (Char/Nchar/Text/Ntext) to Date/time/Datetime/Datetimeoffset/Datetime2/smalldatetime as well?

manisha-deshpande replied:
The current PR handles empty string to date/time conversions at:
1. tsql_coerce_string_literal_hook
2. varchar2date/varchar2time functions

I've added test cases demonstrating the current limitations(`error: "invalid input syntax for type date: """`) with empty char/nchar (~will add text/ntext[TBD: @manisha-deshpande]~ DONE) in `date_time_empty_input_tests.out` below the comment **BABEL-5923**, a new clone jira that tracks broader whitespace handling (including TAB) for all datetime types.

The limitation is stemming from implementing the fix in the extension layer rather than in PostgreSQL's core `date_in/time_in` functions. Since char/nchar/text/ntext conversions bypass the hook (UNKNOWNOID check) and through find_coercion_pathway, directly reach these core functions that don't have the empty string handling.

Current support for empty inputs is limited to conversions from:
- String literals
- varchar/nvarchar
- datetime datatypes

We can track support for this specific issue in BABEL-5923. If this case really needs to be handled soon, we could revert the changes to move to date_in/time_in or brainstorm an alternate fix to handle this. What do you suggest?

Anikait143 replied:
It's fine if we create a Jira to handle cases for empty string (Char/Nchar/Text/Ntext) to Date/time/Datetime/Datetimeoffset/Datetime2/smalldatetime for now. We can revisit this later.

manisha-deshpande replied:
BABEL-5923 will track both:
- TAB space input handling for all date time datatypes (date, time, datetime, smalldatetime, datetime2, datetimeoffset)
- Empty,Whitespace,TAB space char/nchar/text/ntext input handling for date and time.
since a common solution would resolve both issues. 

The end of the file [test/JDBC/expected/date_time_empty_input_tests-vu-verify.sql](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3807/files#diff-b40cf1c3f2d8b263d2882611560dc2c9b321645b60b3ed04c8e2115e3d8f2a8d) tracks the above unsupported test cases.

Anikait143 reviewed (changes requested):
Please sync the branch with upstream we have new tests added which might need some of the fixing

manisha-deshpande committed:
Renamed test file

Task: BABEL-3433
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

manisha-deshpande commented:
> Please sync the branch with upstream we have new tests added which might need some of the fixing

Already updated results in: `Date.out`, `Time.out`
Note: A new (flaky) test is failing in `Datetime_system_functions` but the change is updated by @herambh-shah in a different PR https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3817/files#r2154941034
cc: @Anikait143

KushaalShroff commented on contrib/babelfishpg_common/src/datetime.c:
```diff
@@ -1201,6 +1201,58 @@ initializeToDefaultDatetime(void)
+	if (!str)
+		return true;
+
+	while (*str)
+	{
+		if (!isspace((unsigned char)*str))
+			return false;
+		str++;
+	}
```
In C wont this have potential to cross string boundaries?

Anikait143 replied:
+1, We should have a string length check as well to avoid reading extra memory data.

manisha-deshpande replied:
fixed

KushaalShroff commented on contrib/babelfishpg_tsql/src/pltsql_coerce.c:
```diff
@@ -2566,11 +2568,40 @@ tsql_coerce_string_literal_hook(Oid targetTypeId,
+		{
+			/*
+			 * Set input to default '1900-01-01' if empty string encountered
+			 */
+			DateADT date;
+			struct pg_tm tt,
+						 *tm = &tt;
+			tm->tm_year = 1900;
+			tm->tm_mon = 1;
+			tm->tm_mday = 1;
+			date = date2j(tm->tm_year, tm->tm_mon, tm->tm_mday) - POSTGRES_EPOCH_JDATE;
+
+			newcon->constvalue = DateADTGetDatum(date);
```
Arent all of these conditions internal logic duplicate to initializeToDefaultDate/initializeToDefaultTime. If they are, kindly avoid duplication and please use the functions directly

manisha-deshpande replied:
Updated varchar2date/time to use the existing `initializeToDefaultDateTime` followed by adjusting to date/time instead of creating the new functions initializeToDefaultDate/initializeToDefaultTime.

However, in the case of `babelfishpg_tsql/src/pltsql_coerce.c`, retained this duplicate logic, since direct function import errored out as but the logic is kept identical to maintain consistency. 
```
#include "../../babelfishpg_common/src/datetime.h" failed the babelfish building and final compile step: `...function initializeToDefaultDateTime not recognized in babelfish_tsql.so...`
```

KushaalShroff replied:
Thats not the right way of importing between extensions. Please check this out: https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/BABEL_5_X_DEV/contrib/babelfishpg_common/src/babelfishpg_common.h#L38
If needed, we can have a discussion about it offline.

manisha-deshpande replied:
Thank you for the reference. Fixed.

manisha-deshpande committed:
Resolved PR comments

Removed new functions defaultdate/time to reuse the existing function
initializeToDefaultDatetime and perform additional adjustments to get
Date and Time return types. Added additional tests.

Task: BABEL-3433
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

KushaalShroff commented on contrib/babelfishpg_common/src/varchar.c:
```diff
@@ -1092,7 +1105,25 @@ varchar2time(PG_FUNCTION_ARGS)
+		if (timestamp2tm(ts, NULL, &tt, &fsec, NULL, NULL) != 0)
+			ereport(ERROR,
+					(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
+					 errmsg("timestamp out of range")));
```
Should the error message indicate this happened during time conversion?

KushaalShroff commented on contrib/babelfishpg_common/src/varchar.c:
```diff
@@ -1068,7 +1068,20 @@ varchar2date(PG_FUNCTION_ARGS)
 	 */
 	if (isEmptyOrWhitespace(str))
-		date = initializeToDefaultDate();
+	{
+		/* Extract date from default timestamp (1900-01-01 00:00:00) */
+		struct pg_tm tt;
+		fsec_t fsec = 0;
+		Timestamp ts = initializeToDefaultDatetime();
+
+		if (timestamp2tm(ts, NULL, &tt, &fsec, NULL, NULL) != 0)
+			ereport(ERROR,
+					(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
+					 errmsg("timestamp out of range")));
+
+		/* Convert pg_tm to DateADT */
+		date = date2j(tt.tm_year, tt.tm_mon, tt.tm_mday) - POSTGRES_EPOCH_JDATE;
+	}
```
We could have still used the initializeToDefaultDate in common extension, right?
And using the plugin_ptr we can expose it to TSQL Extension as well.

KushaalShroff commented on contrib/babelfishpg_common/src/varchar.c:
```diff
@@ -1092,7 +1105,25 @@ varchar2time(PG_FUNCTION_ARGS)
+		/* Extract time from default timestamp (1900-01-01 00:00:00) */
+		struct pg_tm tt;
+		fsec_t fsec = 0;
+		Timestamp ts = initializeToDefaultDatetime();
```
In you earlier revision for initializeToDefaultTime you set hr, min & sec to 0 but did not set year/month/day whereas initializeToDefaultDatetime sets it.
```
TimeADT
initializeToDefaultTime(int32 typmod)
{
	TimeADT	time;
	struct pg_tm tt,
				 *tm = &tt;
	tm->tm_hour = tm->tm_min = tm->tm_sec = 0;
	tm2time(tm, 0, &time);
	AdjustTimeForTypmod(&time, typmod);

	return time;
}

/*
 * Set input to default '1900-01-01 00:00:00' if empty string encountered
 */
Timestamp
initializeToDefaultDatetime(void)
{
	Timestamp	result;
	struct pg_tm tt,
			   *tm = &tt;

	tm->tm_year = 1900;
	tm->tm_mon = 1;
	tm->tm_mday = 1;
	tm->tm_hour = tm->tm_min = tm->tm_sec = 0;

	tm2timestamp(tm, 0, NULL, &result);

	return result;
}
```
Can you please clarify if 1900/1/1 is expected and how it was being init earlier?

If I was do something like (pseudo syntax) :
```
Cast(
cast(cast(' ' as varchar) as time )
+ cast(cast('1900/1/2 12:00:00' as varchar)  as time)
as datetime)
```
Does the cast to time lose all of the date part - would it be 1900/1/2 or 1900/1/1?

manisha-deshpande replied:
I was not able to test adding the dates (error:`Operand data type time is invalid for add operator.`
However, tested cast from time to datetime to see that date is overridden with default. 
```
1> select  CAST(CAST(CAST('1900/1/2 00:00:00' AS VARCHAR) AS TIME) AS DATETIME) AS varchar_to_time_to_datetime;
2> go
varchar_to_time_to_datetime
---------------------------
    1900-01-01 00:00:00.000

(1 rows affected)
1> select  CAST(CAST(CAST('' AS VARCHAR) AS TIME) AS DATETIME) AS varchar_to_time_to_datetime;
2> go
varchar_to_time_to_datetime
---------------------------
    1900-01-01 00:00:00.000

(1 rows affected)
```
Found that it is already handled by internal calls to predefined _cast functions_ like `time_datetime(/2/offset,...)` (`sys.time2datetime`) which seem to also set the default date for functions like `DATEADD`.
```
/* time_datetime()
 * Convert time to datetime
 */
Datum
time_datetime(PG_FUNCTION_ARGS)
{
	TimeADT		timeVal = PG_GETARG_TIMEADT(0);
	Timestamp	result;

	struct pg_tm tt,
			   *tm = &tt;
	fsec_t		fsec;

	/* Initialize default year, month, day */
	tm->tm_year = 1900;
	tm->tm_mon = 1;
	tm->tm_mday = 1;

	/* Convert TimeADT type to tm  */
	time2tm(timeVal, tm, &fsec);

	if (tm2timestamp(tm, fsec, NULL, &result) != 0)
		ereport(ERROR,
				(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
				 errmsg("data out of range for datetime")));

	PG_RETURN_TIMESTAMP(result);
}
```
`/* babelfish_extensions/contrib/babelfishpg_common/src/datetime.c */`

KushaalShroff replied:
Great! Thank you for checking it thoroughly

KushaalShroff commented on contrib/babelfishpg_tsql/src/pltsql_coerce.c:
```diff
@@ -2741,6 +2712,33 @@ tsql_coerce_string_literal_hook(Oid targetTypeId,
+					}
+				case TIMEOID:
+					{
+						/* Set input to default '00:00:00' for empty strings */
+						struct pg_tm tt, *tm = &tt;
+						TimeADT time;
+
+						tm->tm_hour = tm->tm_min = tm->tm_sec = 0;
+						tm2time(tm, 0, &time);
+						AdjustTimeForTypmod(&time, targetTypeMod);
+
+						newcon->constvalue = TimeADTGetDatum(time);
```
Referring to my earlier comment: Do we need to init the date part?

manisha-deshpande replied:
As mentioned in my [reply](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3807#discussion_r2164874617) above, we will simply handle only timestamp in the default time function and _don't_ need to set the date part explicitly. 
Any further casts to datetime from time will be handled by cast functions (`time_datetime()`) which sets the default date for time datatype.

manisha-deshpande committed:
Introduced default date time functions in common utility plugin

Task: BABEL-3433
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

manisha-deshpande committed:
Updated failing test cases in other files

Task: BABEL-3433
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>

RIC06X commented:
Overall LGTM, please fix the following error:
```
2025-06-24 21:37:14,689-INFO-****************************** Connection closed to localhost Server successfully!! ******************************
2025-06-24 21:37:14,691-INFO-
--- /home/runner/work/babelfish_extensions/babelfish_extensions/test/python/expected/upgrade_validation/expected_dependency.out	2025-06-24 21:17:28.478550252 +0000
+++ /home/runner/work/babelfish_extensions/babelfish_extensions/test/python/output/upgrade_validation/expected_dependency.out	2025-06-24 21:37:14.687651758 +0000
@@ -746,7 +746,6 @@
 Function sys.varbinaryrowversion(sys.bbf_varbinary,integer,boolean)
 Function sys.varbinaryvarchar(sys.bbf_varbinary,integer,boolean)
 Function sys.varchar2bit(sys."varchar")
-Function sys.varchar2date(sys."varchar")
 Function sys.varchar2datetime(character varying)
 Function sys.varchar2datetime2(character varying)
 Function sys.varchar2datetime2(sys."varchar")

2025-06-24 21:37:14,691-ERROR-Test Failed!
Traceback (most recent call last):
  File "/home/runner/work/babelfish_extensions/babelfish_extensions/test/python/upgrade_validation.py", line 349, in <module>
    main()
  File "/home/runner/work/babelfish_extensions/babelfish_extensions/test/python/upgrade_validation.py", line 346, in main
    assert result == True
AssertionError
```

manisha-deshpande committed:
Remove sys.varchar2date expected dependency from upgrade_validation

Task: BABEL-3433
Signed-off-by: Manisha Deshpande <mmdeshp@amazon.com>
